### DISCUSSION TITLE
How to ensure a rotation is never halted

### DISCUSSION THREAD

**Discussion by BrianCollinss** [APSIMInitiative/ApsimX#9108] [Q&A] [answered]
![image](https://github.com/APSIMInitiative/ApsimX/assets/39573622/71dd8f0f-88f3-4529-b0fb-c903ca92529c)

In this example rotation (from Examples/Rotation.apsimx), the trick currently adopted to ensure simulations are not ever halter is to force Maize and Wheat to be planted at the end of the planting window whether the planting conditions are met to not (i.e., 'Must sow' box is checked). 

![image](https://github.com/APSIMInitiative/ApsimX/assets/39573622/9a9a3794-e6d5-4ea3-a33c-cbf2691b3c28)

What is the best approach to simulating the situation when Maize is not planted in some seasons, but rotation continues into the future? I understand that `[MaizeManager].Script.CanSow` can be modified to make this happen, or Maize can be planted and killed immediately. But, is there a better approach? Should I add a new path from Summer Fallow to Initial (parallel to the one that passes through Maize) that is used when Maize is failed to be planted?

**Comment by BrianCollinss** [comment]
@peter-devoil could you please help me with this?

**Comment by peter-devoil** [comment]
Indeed there are always alternatives - plasticity is an important feature of farming systems. The key is to identify a) what indicates we're past the end of the planting window (ie the rule), and b) what would the farmer do (the action)? 

a) is a simple test: whether the clock.Today is the day after the sowing window closes:
![image](https://github.com/APSIMInitiative/ApsimX/assets/1924386/0ab1d632-cfb5-46fe-9c32-fae0fe2ecfde)

b) is beyond what I can say: I'd guess in this trivial example they'd go to the "Initial" state and wait for a wheat opportunity:
![image](https://github.com/APSIMInitiative/ApsimX/assets/1924386/7a9fd35e-f5f4-4f0c-a2e8-0c4096608dba)

**Reply by BrianCollinss** [reply]
I already tried this, but got this error:
`Error while evaluating transition from Summer Fallow to Initial - rule '': Test condition returned nothing`

**Reply by BrianCollinss** [reply]
![image](https://github.com/APSIMInitiative/ApsimX/assets/39573622/3f75ff07-56bd-4edf-8cc4-25d6421e7bce)

**Comment by peter-devoil** [comment, marked as answer]
Several buglets have slipped into the rotation/bubble chart (@par456 - is there still a list of these somewhere?):

- blank lines show up in the rules/conditions and it tries to evaluate them
- the arc selection procedure (select the destination 1st, then right click on the source) doesnt match the instruction (select the source node 1st, then right click on the destination)
- when switching away / returning to the bubble chart component, a previously selected arc remains highlighted, but the selected condition/action windows arent populated on return.
- The textbox selection colours are invisible in dark mode

You've got the 1st: delete the whitespace at the end of the list of conditions.

**Reply by par456** [reply]
We don't have a list, but these will be all things I introduced when I overhauled the gui side of it a few months ago. I'll do up an issue with those listed, I know where in the GUI code the problem will need to be fixed.

**Reply by BrianCollinss** [reply]
Thanks Peter. It seems it is now doing what it is supposed to do:

![image](https://github.com/APSIMInitiative/ApsimX/assets/39573622/7132761b-7c62-4589-8cc2-52ab33bfc776)

Strange things is that I just needed to remove that whitespace once. Next time, it runs without my needing to remove it!

Also I think there is an issue in the way states are reported: https://github.com/APSIMInitiative/ApsimX/issues/9111. Ideally, I would like to see Wheat, Initial, Summer Fallow, ... in the last column.

**Comment by BrianCollinss** [comment]
This bit:

```
        [Units("0-1")]
        public int NotPlanted
        {
        	get 
        	{
        		if (!afterInit)
                    return 0;
        		var dayAfter = DateUtilities.GetDate(EndDate, Clock.Today.Year).AddDays(1);		
        		return((Clock.Today >= dayAfter) ? 1 : 0);
        	}
        }
```

is run even when it is Wheat's turn to be planted. Therefore, it is always set at 1 in the middle of Wheat season. Hence, Maize never gets to be planted. I tried another approach as well, but all I can think of requires the managers to be aware of what crop is or will be on the ground, or at least have access to the parameters (e.g., planting window) of the other manager.

![image](https://github.com/APSIMInitiative/ApsimX/assets/39573622/374d820c-83d4-433d-90a1-55c813e22be9)

**Comment by peter-devoil** [comment]
It's only called when asked: so if you call it during the Summer fallow, it must have been preceded by the wheat crop. And if there is never an opportunity to sow maize (eg very restrictive conditions), you will always have a wheat crop. That's the intent?

**Reply by BrianCollinss** [reply]
In this case, I tested it by forcing Maize to be planted at the end of the window. The issue is that that piece of code is run every day. I need to find a way to have it run just after wheat is harvested.

**Comment by peter-devoil** [comment]
The code is called - it doesnt subscribe to any events, and can only be run by a method call from another component. That component is the rotation manager, which has it as a condition to change state from A to B. If the current state is C, then it wont be called. If the current state is A, it will be called (along with any other conditions leading out of A). 

The rug plot component(when it's working) shows the results of these calls (1,0 ...) in the bottom right hand corner. Clicking on a day in the rug shows the rules/conditions evaluated on that day

**Reply by BrianCollinss** [reply]
I understand that. What I'm dealing with is that that code has been run before by the Maize manager and has already been set at 1 before maize planting window is reached. So when the time comes to evaluate maize planting requirements, that flag is already 1, keeping maize from being planted.

**Reply by peter-devoil** [reply]
Nothing should be calling it in the maize manager. It's defined, but not called. It's a get() method, so only runs when called.

**Reply by BrianCollinss** [reply]
Ah! I get it. It is run by the report!!

![image](https://github.com/APSIMInitiative/ApsimX/assets/39573622/0ce7da6f-5de6-4c31-94e5-2ff67765356e)

**Reply by peter-devoil** [reply]
You have to be careful reading its value from the report; as reporting occurs at the end of the day, well after any management is done - so you're looking at the value after the fact, not before. That's one of the reasons the rugplot was developed - to record the management steps - there may be several transitions within one day.

**Reply by BrianCollinss** [reply]
Still does not plant Maize. Wheat is harvested late-August to early-September. Then, state is Summer Fallow and RotationManager starts calling that property. As maize planting window is April-Jun (in this example), that flag will be set at 1, and maize is bypassed.

**Reply by BrianCollinss** [reply]
I fixed it like this (checking if yesterday was within Maize planting window):

```
        [Units("0-1")]
        public int NotPlanted
        {
        	get 
        	{
        		if (!afterInit)
                    return 0;
        		var dayAfter = DateUtilities.GetDate(EndDate, Clock.Today.Year).AddDays(1);	
        		bool inWindow = DateUtilities.WithinDates(StartDate, Clock.Today.AddDays(-1), EndDate);	
        		return((Clock.Today >= dayAfter && inWindow) ? 1 : 0);
        	}
        }
```
